rust/src/storage/s3/dynamodb_lock.rs

```diff
@@ -0,0 +1,478 @@
+        }
+    }
+}
+
+impl From<RusotoError<GetItemError>> for DynamoError {
+    fn from(error: RusotoError<GetItemError>) -> Self {
+        match error {
+            RusotoError::Service(GetItemError::ResourceNotFound(_)) => DynamoError::TableNotFound,
+            _ => DynamoError::GetItemError(error),
+        }
+    }
+}
+/// The partition key field name in DynamoDB
+pub const PARTITION_KEY_NAME: &str = "key";
+/// The field name of `owner_name` in DynamoDB
+pub const OWNER_NAME: &str = "ownerName";
+/// The field name of `record_version_number` in DynamoDB
+pub const RECORD_VERSION_NUMBER: &str = "recordVersionNumber";
+/// The field name of `is_released` in DynamoDB
+pub const IS_RELEASED: &str = "isReleased";
+/// The field name of `lease_duration` in DynamoDB
+pub const LEASE_DURATION: &str = "leaseDuration";
+/// The field name of `data` in DynamoDB
+pub const DATA: &str = "data";
+
+mod expressions {
+    /// The expression that checks whether the lock record does not exists.
+    pub const ACQUIRE_LOCK_THAT_DOESNT_EXIST: &str = "attribute_not_exists(#pk)";
+
+    /// The expression that checks whether the lock record exists and it is marked as released.
+    pub const PK_EXISTS_AND_IS_RELEASED: &str = "attribute_exists(#pk) AND #ir = :ir";
+
+    /// The expression that checks whether the lock record exists
+    /// and its record version number matches with the given one.
+    pub const PK_EXISTS_AND_RVN_MATCHES: &str = "attribute_exists(#pk) AND #rvn = :rvn";
+
+    /// The expression that checks whether the lock record exists,
+    /// its record version number matches with the given one
+    /// and its owner name matches with the given one.
+    pub const PK_EXISTS_AND_OWNER_RVN_MATCHES: &str =
+        "attribute_exists(#pk) AND #rvn = :rvn AND #on = :on";
+
+    /// The expression that updates is_released and data fields of DynamoDB record.
+    pub const UPDATE_IS_RELEASED_AND_DATA: &str = "SET #ir = :ir, #d = :d";
+
+    /// The expression that updates is_released field of DynamoDB record.
+    pub const UPDATE_IS_RELEASED: &str = "SET #ir = :ir";
+}
+
+mod vars {
+    pub const PK_PATH: &str = "#pk";
+    pub const RVN_PATH: &str = "#rvn";
+    pub const RVN_VALUE: &str = ":rvn";
+    pub const IS_RELEASED_PATH: &str = "#ir";
+    pub const IS_RELEASED_VALUE: &str = ":ir";
+    pub const OWNER_NAME_PATH: &str = "#on";
+    pub const OWNER_NAME_VALUE: &str = ":on";
+    pub const DATA_PATH: &str = "#d";
+    pub const DATA_VALUE: &str = ":d";
+}
+
+/// Provides a simple library for using DynamoDB's consistent read/write feature
+/// to use it for managing distributed locks.
+pub struct DynamoDbLockClient {
+    client: DynamoDbClient,
+    opts: Options,
+}
+
+impl DynamoDbLockClient {
+    /// Creates new DynamoDB lock client
+    pub fn new(client: DynamoDbClient, opts: Options) -> Self {
+        Self { client, opts }
+    }
+
+    /// Attempts to acquire a lock until it either acquires the lock or the error is found.
```

Comment:
could you expand a little bit more on what are the categories of errors that would break the loop?

Reply:
I'll rewrite a documentation once change the return type to Option<LockItem> so it'll be 1:1 with original's